# Individual Planning Stage (DSCI 100 Group Project)
Student Name: Vicky Tan

Student Number: 97847115

In [51]:
library(tidyverse)
library(lubridate)

## 1. Data Description

### Dataset #1: players.csv

In [52]:
url_players <- "https://raw.githubusercontent.com/vckytn22/DSCI-100-004-27-Individual-Planning-Stage-Vicky-Tan/refs/heads/main/players.csv"
players_data <- read_csv(url_players)
head(players_data)

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


experience subscribe
1 Pro         TRUE    
2 Veteran     TRUE    
3 Veteran    FALSE    
4 Amateur     TRUE    
5 Regular     TRUE    
6 Amateur     TRUE    
  hashedEmail                                                      played_hours
1 f6daba428a5e19a3d47574858c13550499be23603422e6a0ee9728f8b53e192d 30.3        
2 f3c813577c458ba0dfef80996f8f32c93b6e8af1fa939732842f2312358a88e9  3.8        
3 b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3c5a9d2118eb7ccbb28  0.0        
4 23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4fa7a5a659ff443a0eb5  0.7        
5 7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb0af4d48fcce2420f3e  0.1        
6 f58aad5996a435f16b0284a3b267f973f9af99e7a89bee0430055a44fa92f977  0.0        
  name      gender Age
1 Morgan    Male    9 
2 Christian Male   17 
3 Blake     Male   17 
4 Flora     Female 21 
5 Kylie     Male   21 
6 Adrian    Female 17

In [53]:
age_summary <- players_data |>
    summarize(
        mean_age = round(mean(Age, na.rm = TRUE), 2),
        median_age = round(median(Age, na.rm = TRUE), 2),
        standard_deviation_age = round(sd(Age, na.rm = TRUE), 2),
        min_age = round(min(Age, na.rm = TRUE), 2),
        max_age = round(max(Age, na.rm = TRUE), 2))
age_summary

mean_age median_age standard_deviation_age min_age max_age
1 21.14    19         7.39                   9       58

In [54]:
played_hours_summary <- players_data |>
    summarize(
        mean_played_hours = round(mean(played_hours, na.rm = TRUE), 2),
        median_played_hours = round(median(played_hours, na.rm = TRUE), 2),
        standard_deviation_played_hours = round(sd(played_hours, na.rm = TRUE), 2),
        min_played_hours = round(min(played_hours, na.rm = TRUE), 2),
        max_played_hours = round(max(played_hours, na.rm = TRUE), 2))
played_hours_summary

mean_played_hours median_played_hours standard_deviation_played_hours
1 5.85              0.1                 28.36                          
  min_played_hours max_played_hours
1 0                223.1

### Summary Statistics for Dataset #1
Origin
- dataset details the identity of players was collected through the recording of players' actions in Minecraft servers
- collected by Pacific Laboratory of Artificial Intelligence (PLAI), a research group in Computer Science at the University of British Columbia led by Frank Wood
  
Overall Summary
- 196 observations
- 7 variables

Name of Variables and Types
- experience is character type, meaning it represents text values
- subscribe is logical type, meaning it contains boolean values, TRUE or FALSE, for each corresponding observation
- hashedEmail is character type, meaning it stores text values
- played_hours is double(numeric) type, meaning it records decimal numbers
- name is character type, meaning it displays text values
- gender is characer type, meaning it shows text values
- Age is double(numeric) type, meaning it shows numbers and in this case integers

Variable Description
- experience is the level of expertise
- subscribe is whether or not the player has a subcription
- hashedEmail is their email identifier
- played_hours is the number of hours spent on the game
- name is the player's name
- gender is the player's gender
- Age is the player's age in years

Numerical Summary of played_hours in Table Below

In [55]:
played_hours_summary <- players_data |>
    summarize(
        mean_played_hours = round(mean(played_hours, na.rm = TRUE), 2),
        median_played_hours = round(median(played_hours, na.rm = TRUE), 2),
        standard_deviation_played_hours = round(sd(played_hours, na.rm = TRUE), 2),
        min_played_hours = round(min(played_hours, na.rm = TRUE), 2),
        max_played_hours = round(max(played_hours, na.rm = TRUE), 2))
played_hours_summary

mean_played_hours median_played_hours standard_deviation_played_hours
1 5.85              0.1                 28.36                          
  min_played_hours max_played_hours
1 0                223.1

Numerical Summary of Age in Table Below

In [56]:
age_summary <- players_data |>
    summarize(
        mean_age = round(mean(Age, na.rm = TRUE), 2),
        median_age = round(median(Age, na.rm = TRUE), 2),
        standard_deviation_age = round(sd(Age, na.rm = TRUE), 2),
        min_age = min(Age, na.rm = TRUE),
        max_age = max(Age, na.rm = TRUE))
age_summary

mean_age median_age standard_deviation_age min_age max_age
1 21.14    19         7.39                   9       58

Potential Issues
- median_played_hours is 0.1, suggesting the majority of observations report 0 played hours
  - may present some difficulty in analyzing played_hours as the data is highly skewed towards no play time, which results in limited variation

### Dataset #2: sessions.csv

In [57]:
url_sessions <- "https://raw.githubusercontent.com/vckytn22/DSCI-100-004-27-Individual-Planning-Stage-Vicky-Tan/refs/heads/main/sessions.csv"
sessions_data <- read_csv(url_sessions)
head(sessions_data)

Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


hashedEmail                                                     
1 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
2 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
3 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
4 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
5 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
6 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
  start_time       end_time         original_start_time original_end_time
1 30/06/2024 18:12 30/06/2024 18:24 1.71977e+12         1.71977e+12      
2 17/06/2024 23:33 17/06/2024 23:46 1.71867e+12         1.71867e+12      
3 25/07/2024 17:34 25/07/2024 17:57 1.72193e+12         1.72193e+12      
4 25/07/2024 03:22 25/07/2024 03:58 1.72188e+12         1.72188e+12      
5 25/05/2024 16:01 25/05/2024 16:12 1.71665e+12         1.71665e+12      
6 23/06/2024 15:08 23/06/2024 17:10 1.71916e+12         1.71916e+12

### Summary Statistics for Dataset #2

Origin
- dataset records the exact start and end times, including dates, of each player's gaming session
  - formatted in DD/MM/YYYY HH/MM and UNIX time
- collected by Pacific Laboratory of Artificial Intelligence (PLAI), a research group in Computer Science at the University of British Columbia led by Frank Wood
  
Overall Summary
- 1535 observations
- 5 variables

Name of Variables and Types
- hashedEmail is character type, meaning it stores text values
- start_time and end_time are character types, meaning they contain text values
  - however, they can be converted into POSIXct, a date-time class in R
  - as.Date and as.hour functions can also be used to convert data for more time focussed analysis
- original_start_time and original_end_time are double (numeric type), meaning numerals that represent the miliseconds

Variable Description
- hashedEmail is their email identifier
- start_time is the time when the player started gaming, down to the exact minute of the date
- end_time is the time when the player stopped gaming, down to the exact minute of the date
- original_start_time is the game's start in UNIX format
- original_end_time is the game's end in UNIX format

Numerical Summary of Session Length in Table Below
- start_time and end_time were converted into POSIXct format
- session length is double(numeric) type, representing the gaming duration of one session in minutes
    - this variable shows how long players game per session
    - this was done because calculating the summary statistics of the time of day or date in the year is not straightforward and entirely logical

In [60]:
sessions_data_converted <- sessions_data |>
    mutate(
    start_time = dmy_hm(start_time),
    end_time = dmy_hm(end_time))
session_length_data <- sessions_data_converted |>
    mutate(session_length = as.numeric(difftime(end_time, start_time, units = "mins")))

session_length_summary <- session_length_data |>
    summarize(
        mean_session_length = round(mean(session_length, na.rm = TRUE), 2),
        median_session_length = round(median(session_length, na.rm = TRUE), 2),
        standard_session_length = round(sd(session_length, na.rm = TRUE), 2),
        min_session_length = min(session_length, na.rm = TRUE),
        max_session_length = max(session_length, na.rm = TRUE))
session_length_summary

mean_session_length median_session_length standard_session_length
1 50.86               30                    55.57                  
  min_session_length max_session_length
1 3                  259

## 2. Questions

Broad Question: What player characteristics and behaviours are most predictive of subscribing to a game-related newsletter, and how do these features differ between various player types?

Specific Question: Can played_hours, age, and experience predict if a player is subscribed to the game? 



How the Data will Address the Question of Interest
- I will be utilizing variables from both datasets to provide a broader perspective through multiple variables.
- Session length is indicative of the player's liking of the game and their engagement habits.
    - Longer session lengths should correspond to a higher chance of subscription.
- Age can influence a player's financial ability to subscribe to the game, the amount of time they can devote to playing, and their likelihood of enjoying it, as Minecraft's primary target audience is children and teenagers.
    - Thus, several insights on subscription can be made with age.
- Experience, though subjective, can inform on skill and engagement level.
    - More skillful players, Pro or Veteran, have a higher chance of being subscribed.

Wrangling Data
- I plan organize my three variables in a single tidy data table.
- Session length has already been transformed from start and end time.
- Thus, session length and age are numeric predictors that can be used in K-nearest-neighbours (KNN) classification.
- The experience variable will be explored through bar plots.

DO I HAVE TO JUSTIFY DROPPING CERTAIN VARIABLES

## 3. Exploratory Data Analysis and Description

In this assignment, you will:

Demonstrate that the dataset can be loaded into R.
Do the minimum necessary wrangling to turn your data into a tidy format. Do not do any additional wrangling here; that will happen later during the group project phase.
Compute the mean value for each quantitative variable in the players.csv data set. Report the mean values in a table format.
Make a few exploratory visualizations of the data to help you understand it.
Use our visualization best practices to make high-quality plots (make sure to include labels, titles, units of measurement, etc)
Explain any insights you gain from these plots that are relevant to address your question
Note: do not perform any predictive analysis here. We are asking for an exploration of the relevant variables to demonstrate that you understand them well before performing any additional modelling, and to identify potential problems you anticipate encountering.

The two datasets have already been loaded into R in section 1, data description. This was done because it was necessary to determine their numerical summary statistics. 

Wrangling Data into Tidy Format Below

- Thought Process for Session Length Data Wrangling
    - The table session_length_data records each gaming session as a new observation, thus repeating the same player information.
    - KNN classification would process the same player as multiple different ones, biasing predictions toward players with more sessions
    - To account for this the average session length for each player will be used instead

Thought Process for Finalized Data 
- Joining two datasets, players_data and mean_session_data, by matching players' hashed emails
    - ensures that session length corresponds to the correct subscription status
- selecting the desired predictor variables
- 196 observations indicating 196 different players

In [84]:
mean_session_data <- session_length_data |>
    group_by(hashedEmail) |>
    summarize(mean_session = mean(session_length, na.rm = TRUE))

subscription_data <- players_data |>
    left_join(mean_session_data, by = "hashedEmail") |>
    select(subscribe, experience, Age, mean_session)
head(subscription_data)

subscribe experience Age mean_session
1  TRUE     Pro         9  74.77778    
2  TRUE     Veteran    17  85.00000    
3 FALSE     Veteran    17   5.00000    
4  TRUE     Amateur    21  50.00000    
5  TRUE     Regular    21   9.00000    
6  TRUE     Amateur    17        NA

- histogram of mean_session, colour coded with subscribe, most people
- age versus experience, does age depend on experience, colour coded with subscribe, bar plot
- scatter of age versus mean_session
- mean_session and experience bar plot